runtime-modules/referendum/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+    fn vote(
+        account_id: &<T as system::Trait>::AccountId,
+        commitment: &T::Hash,
+        balance: &Balance<T, I>,
+    ) -> Result<(), Error<T, I>> {
+        // lock stake amount
+        T::Currency::set_lock(
+            T::LockId::get(),
+            account_id,
+            *balance,
+            WithdrawReason::Transfer.into(),
+        );
+
+        // store vote
+        Votes::<T, I>::insert(
+            account_id,
+            SealedVote {
+                commitment: *commitment,
+                balance: *balance,
+                cycle_id: CurrentCycle::<I>::get(),
+                vote_for: None,
+            },
+        );
+
+        Ok(())
+    }
+
+    fn reveal_vote(
+        account_id: &<T as system::Trait>::AccountId,
+        option_index: &u64,
+        sealed_vote: &SealedVote<T::Hash, Balance<T, I>>,
+    ) -> Result<(), Error<T, I>> {
+        let distribute_vote =
+            |stage_data: &mut ReferendumStageRevealing<T::BlockNumber, T::VotePower>| {
+                // calculate vote power
+                let vote_power = T::caclulate_vote_power(account_id, &sealed_vote.balance);
+                stage_data.revealed_votes[*option_index as usize] += vote_power;
+            };
+
+        // store revealed vote
+        Stage::<T, I>::mutate(|stage| {
+            match stage {
+                ReferendumStage::Revealing(stage_data) => distribute_vote(stage_data),
+                _ => panic!("Invalid state"), // will never happen
+            }
+        });
+
+        // remove user commitment to prevent repeated revealing
+        Votes::<T, I>::mutate(account_id, |vote| (*vote).vote_for = Some(*option_index));
+
+        Ok(())
+    }
+
+    fn release_stake(account_id: &<T as system::Trait>::AccountId) {
+        // lock stake amount
+        T::Currency::remove_lock(T::LockId::get(), account_id);
+
+        // remove vote record
+        Votes::<T, I>::remove(account_id);
+    }
+}
+
+/////////////////// Ensure checks //////////////////////////////////////////////
+
+struct EnsureChecks<T: Trait<I>, I: Instance> {
+    _dummy: PhantomData<(T, I)>, // 0-sized data meant only to bound generic parameters
+}
+
+impl<T: Trait<I>, I: Instance> EnsureChecks<T, I> {
+    /////////////////// Common checks //////////////////////////////////////////
+
+    fn ensure_super_user(origin: T::Origin) -> Result<T::AccountId, Error<T, I>> {
+        let account_id = ensure_signed(origin)?;
+
+        // ensure superuser requested action
+        if !T::is_super_user(&account_id) {
+            return Err(Error::OriginNotSuperUser);
+        }
+
+        Ok(account_id)
+    }
+
+    fn ensure_regular_user(origin: T::Origin) -> Result<T::AccountId, Error<T, I>> {
+        let account_id = ensure_signed(origin)?;
+
+        Ok(account_id)
+    }
+
+    /////////////////// Action checks //////////////////////////////////////////
+
+    fn can_start_referendum_extrinsic(
```

Comment:
I think `can_start_referendum_extrinsic` can in its entirety be replaced by a runtime `EnsureOrigin` runtime trait. this is the typical way of making extrinsic access customizable. See here for example

https://github.com/paritytech/substrate/blob/master/frame/treasury/src/lib.rs#L146

The reason we do not use it alot is that we often want to to authenticate a memberid + origin together.

In the scheme you have currently implemented, the origin _must_. be an account, but this will generally not be the case, in general it will be some other part of the runtime, win which case the origin needs to be checked to be a `RootOrigin`.

Reply:
Done